### PR TITLE
ddl: fix bug when add expression index

### LINKED ISSUE #17111: Expression index returns `Cancelled DDL job` error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```
mysql> create table t1 (a varchar(10), b varchar(10)) ;
Query OK, 0 rows affected (0.01 sec)
mysql> alter table t1 add unique index ei_ab ((concat(a, b)));
ERROR 8214 (HY000): Cancelled DDL job
```

### 2. What did you expect to see?

I'm not sure this statement is illegal or not, but TiDB should not throw a `Cancelled DDL job` error anyway.

### 3. What did you see instead?

Either no error or a better error message.

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

master

commit 014d101a5d7fbd78f4bc306251181c01c0f01c7c
Author: Mingcong Han <hanmingcong123@hotmail.com>
Date:   Thu May 7 14:27:05 2020 +0800

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve?

Issue Number: close #17111  

Problem Summary:

```
mysql> create table t1 (a varchar(10), b varchar(10)) ;
Query OK, 0 rows affected (0.01 sec)
mysql> alter table t1 add unique index ei_ab ((concat(a, b)));
ERROR 8214 (HY000): Cancelled DDL job
```

### What is changed and how it works?

What's Changed:

1. Check `Column != nil` before use it and check its flag.
2. Avoid expression index be select as implicit primary key.

> For what is implicit primary key, see the comments before function `getPrimaryKey()` in this PR.

How it works:

In every status change of add index DLL job, it will call `getPrimaryKey()` to check whether the PK is invisible. But, when the index has a column in `DeleteOnly` state,
```
col := model.FindColumnInfo(tblInfo.Cols(), idxCol.Name.L)
if !mysql.HasNotNullFlag(col.Flag) {
	...
}
```
`model.FindColumnInfo()` returns `nil`, we should check if `col` is `nil` before use it.

Another Improvement is we should **NOT** promote expression index as the implicit primary key, as it cannot be primary key.

### Related changes

- None

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- None

### Release note 
- No release note.

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1139,6 +1140,12 @@ func getPrimaryKey(tblInfo *model.TableInfo) *model.IndexInfo {
 			allColNotNull := true
 			for _, idxCol := range key.Columns {
 				col := model.FindColumnInfo(tblInfo.Cols(), idxCol.Name.L)
+				// This Index has a column in DeleteOnly state,
```
Comment (zimulala) [inline ddl/ddl_api.go:1143]:
```suggestion
				// This index has a column in DeleteOnly state,
```

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -1139,6 +1140,12 @@ func getPrimaryKey(tblInfo *model.TableInfo) *model.IndexInfo {
...
+				// or it is expression index (it defined on a hidden column),
+				// it can not be implicit PK, go to next index iterator
+				if col == nil || col.Hidden {
+					continue IndicesLoop
```
Comment (zimulala) [inline ddl/ddl_api.go:1147]: Could we use `break` instead of this line?

Comment (Deardrops) [reply ddl/ddl_api.go:1151]: No, if break here, L1149 will choose it as implicit PK.

Comment (AilinKid) [reply ddl/ddl_api.go:1151]: maybe sister's meaning: use a flag like `allColNotNull`, set it false, break, and check to skip in L1158

File: ddl/index.go [hunk 3]
```diff
@@ -483,6 +483,11 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
 		indexInfo.Unique = unique
 		indexInfo.ID = allocateIndexID(tblInfo)
 		tblInfo.Indices = append(tblInfo.Indices, indexInfo)
+
+		if err := checkInvisibleIndexOnPK(tblInfo); err != nil {
```
Comment (AilinKid) [inline ddl/index.go:487]: add some comment about this?

File: ddl/index.go [hunk 4]
```diff
@@ -483,6 +483,11 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
...
+			job.State = model.JobStateCancelled
```
Comment (djshow832) [inline ddl/index.go:488]: No test case for it?

Comment (Deardrops) [reply ddl/index.go:488]: tested in https://github.com/pingcap/tidb/pull/17151/files#diff-703ae6b7872b425273d1832c198598c8R2100,
`alter table add index` is same with `create index on table`

File: ddl/ddl_api.go (hunk 2 above)
Comment (Deardrops) [reply ddl/ddl_api.go:1151]: > 
> 
> maybe sister's meaning: use a flag like `allColNotNull`, set it false, break, and check to skip in L1158

I think it is same, if there are not any performance and safety problem, I'd like keep it.

Comment (AilinKid) [reply ddl/ddl_api.go:1151]: it's ok

File: ddl/index.go (hunk 3 above)
Comment (Deardrops) [reply ddl/index.go:490]: Done, PTAL @AilinKid

File: ddl/ddl_api.go (hunk 2 above)
Comment (zimulala) [reply ddl/ddl_api.go:1151]: I think if `IndicesLoop` isn't necessary, we'd better don't use it. PTAL @tiancaiamao

Comment (tiancaiamao) [reply ddl/ddl_api.go:1151]: Can we avoid the unconditional jump here?

> Go To Statement Considered Harmful

-- Edsger Dijkstra [[1]](https://en.wikipedia.org/wiki/Considered_harmful)

Comment (Deardrops) [reply ddl/ddl_api.go]: Done

File: ddl/db_integration_test.go [hunk 5]
```diff
@@ -2094,6 +2094,12 @@ func (s *testIntegrationSuite7) TestAddExpressionIndex(c *C) {
 
 	tk.MustQuery("select * from t;").Check(testkit.Rows("1 2.1"))
 
+	// Issue #17111
+	tk.MustExec("drop table if exists t1, t2, t3")
```
Comment (crazycs520) [inline ddl/db_integration_test.go:2098]:
```suggestion
	tk.MustExec("drop table if exists t1")
```
t2,t3 is useless?

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -1133,16 +1133,32 @@ func getPrimaryKey(tblInfo *model.TableInfo) *model.IndexInfo {
 			// table has explicit primary key
 			return key
 		}
+		// The case index without any columns should never happen, but still do a check here
+		if key.Columns == nil || len(key.Columns) == 0 {
```
Comment (crazycs520) [inline ddl/ddl_api.go:1137]:
```suggestion
		if len(key.Columns) == 0 {
```

only one condition will be ok?

**Comment by Deardrops** [comment]
@crazycs520 Comment addressed, PTAL

**Pull request merged**